dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/DynamicSpinLoopDetection.java

```diff
@@ -72,167 +70,108 @@ private List<LoopData> computeLoopData(Function func, LoopAnalysis loopAnalysis)
-                    // This means the next loop iteration will observe the newly written value,
-                    // hence the loop is not side effect free.
-                    loop.sideEffects.add(cur);
-                } else {
-                    safeRegisters.add(writer.getResultRegister());
-                }
+                loop.globalSideEffects.add(cur);
             }
         } while ((cur = cur.getSuccessor()) != loop.getEnd().getSuccessor());
 
-    }
-
-    private void reduceToDominatingSideEffects(LoopData loop) {
-        if (loop.sideEffects.isEmpty()) {
-            return;
-        }
-
-        final List<Event> sideEffects = loop.sideEffects;
-        final Event start = loop.getStart();
-        final Event end = loop.getEnd();
-        final Predicate<Event> isAlwaysSideEffectful = (e -> e.cfImpliesExec() && sideEffects.contains(e));
-
-        final DominatorTree<Event> preDominatorTree = DominatorAnalysis.computePreDominatorTree(start, end);
-        final DominatorTree<Event> postDominatorTree = DominatorAnalysis.computePostDominatorTree(start, end);
-
-        // (1) Delete all side effects that are on exit paths (those have no dominator in the post-dominator tree
-        // because they are no predecessor of the end of the loop body).
-        final Predicate<Event> isOnExitPath = (e -> postDominatorTree.getImmediateDominator(e) == null);
-        sideEffects.removeIf(isOnExitPath);
-
-        // (2) Check if always side-effect-full at the end of an iteration directly before entering the next one.
-        // This is an approximation: If the end of the iteration is predominated by some side effect
-        // then we always observe side effects.
-        loop.isAlwaysSideEffectful = Streams.stream(preDominatorTree.getDominators(end)).anyMatch(isAlwaysSideEffectful);
-        if (loop.isAlwaysSideEffectful) {
-            return;
-        }
-
-        // (3) Delete all side effects that are dominated by another one
-        // NOTE: This can be implemented more efficiently, but maybe we don't need to.
-        for (int i = sideEffects.size() - 1; i >= 0; i--) {
-            final Event sideEffect = sideEffects.get(i);
-            final Iterable<Event> dominators = Iterables.concat(
-                    preDominatorTree.getStrictDominators(sideEffect),
-                    postDominatorTree.getStrictDominators(sideEffect)
-            );
-            final boolean isDominated = Iterables.tryFind(dominators, isAlwaysSideEffectful::test).isPresent();
-            if (isDominated) {
-                sideEffects.remove(i);
-            }
-        }
+        // Every live register that is written to is a potential local side effect.
+        loop.writtenLiveRegisters.addAll(Sets.intersection(
+                writtenRegisters,
+                liveRegsAna.getLiveRegistersAt(loop.getStart())
+        ));
     }
 
     private void instrumentLoop(LoopData loop) {
```

Comment:
I find it hard to keep track in my head on where all the new events are added. Please add some documentation on how the IR will look like after the instrumentation, something along the lines of [this](https://github.com/hernanponcedeleon/Dat3M/blob/master/dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/NaiveLoopBoundAnnotation.java#L32-L43).

Reply:
Thanks. This made reviewing much easier.